src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryLogic.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+
+/**
+ * Allows configurable statement retry through the use of the 'retryExec' connection property. Each rule read in is
+ * converted to ConfigRetryRule objects, which are stored and referenced during statement retry.
+ *
+ */
+public class ConfigurableRetryLogic {
+    private final static int INTERVAL_BETWEEN_READS_IN_MS = 30000;
+    private final static String DEFAULT_PROPS_FILE = "mssql-jdbc.properties";
+    private static final Lock CRL_LOCK = new ReentrantLock();
+    private static final java.util.logging.Logger CONFIGURABLE_RETRY_LOGGER = java.util.logging.Logger
+            .getLogger("com.microsoft.sqlserver.jdbc.ConfigurableRetryLogic");
+    private static ConfigurableRetryLogic singleInstance;
+    private static long timeLastModified;

```

Comment:
As we are not serializing the calls to the statis ConfigurableRetryLogic static methods (i.e. except for the getInstance() method, all the static class objects are inherently not thread safe. Most Java statements are not atomic. That means that internally even a Java assignment operation internally within the JVM results in several statements that are executed. Because of this it is possible in a multithreaded scenario for there to be a switch in the middle of a Java assignment statement (e.g. hypothetically the Java assignment statement results in 4 separate statements in the JVM) and switches to another thread also performing that same assignment statement resulting in corrupted data for one or possibly both threads.

To address this scenario, Java has added atomic class types for manipulating simple objects in a thread safe manner (i.e. the serialization is implicitly performed in the atomic class and does not need to be explicitly performed in our own code). Please convert these static objects shared between thread to use AtomicInteger, AtomicLong, and AtomicReference. This will ensure that a single Java statement manipulating these will be atomic and cannot be impacted by a switch between threads.

Form more details please refer to: https://naveen-metta.medium.com/atomic-operations-in-java-mastering-thread-safety-and-concurrency-7c3360ec0bc5

Reply:
Fair point, done.